### DISCUSSION TITLE
Scaffolding tool for building operators

### DISCUSSION THREAD

**Discussion by 26tanishabanik** [kube-rs/kube#1039] [Ideas]
### Would you like to work on this feature?

yes

### What problem are you trying to solve?

To have a scaffolding on top of `kube-rs`, so that it can be used a plugin feature or extended capability for developing operators in k8s. This would provide a building base for new operators in Rust

### Describe the solution you'd like

A simple scaffolding built on top of `kube-rs`, which can be used as a `CLI` , to build the initial config files for writing an operator, like, ready-made, reconciler functions, CRD yamls, RBAC config files and so on

### Describe alternatives you've considered

Not yet considered

### Documentation, Adoption, Migration Strategy

A CLI abstraction would be useful for this tool

### Target crate for feature

kube-client

**Comment by clux** [comment]
Hey there. I appreciate the initiative on this! I think scaffolding of kube is **a little early** in-org with some of the changes we have planned (there's a lot more room to simplify the controller code which would reduce the need for scaffolding).

Scaffolding needs to be maintained, and at the moment we have enough work accomplishing the goals we have for kube with the associated [controller-rs](https://github.com/kube-rs/controller-rs) and other tooling in the org with the current bandwidth for our maintainers. We did suggest something similar a year back with templating out an entire repo with `cargo-generate` in #580, but it was again deprioritised (not because it's not worth doing, but because there are a lot of other things more worth doing).

Still, a CLI that can do something like what you suggest would be interesting to see, and it could indeed be useful! However; it probably would not need to live inside `kube-rs/kube` but rather it's own repo using `kube` as a library to do the scaffolding (this also keeps CLI / AST specific dependencies needed outside of kube). Honestly, this feels like a good project to try to prototype **outside the org**. Happy to look at what you can come up with and offer suggestions (or you can ask question in the discussions / discord)!
